stepflow-rs/crates/stepflow-execution/src/debug_executor.rs

```diff
@@ -0,0 +1,1177 @@
+//!
+//! [`DebugExecutor`] provides a queue/eval API for interactive debugging of workflows.
+//! It uses [`ItemState`] for dependency tracking and supports recovery from persistent state.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use bit_set::BitSet;
+use error_stack::ResultExt as _;
+use stepflow_core::BlobId;
+use stepflow_core::status::{StepExecution, StepStatus};
+use stepflow_core::{FlowResult, values::ValueRef, workflow::Flow};
+use stepflow_plugin::ExecutionContext;
+use stepflow_state::StateStore;
+use uuid::Uuid;
+
+use crate::state::ItemState;
+use crate::step::{StepExecutionResult, StepInspection, StepMetadata, execute_step_async};
+use crate::{ExecutionError, Result, StepflowExecutor};
+
+/// Debug executor for step-by-step workflow execution.
+///
+/// Provides the queue/eval API for interactive debugging:
+/// - `queue_step()` / `queue_must_execute()` - add steps to execution queue
+/// - `run_next_step()` / `run_queue()` - execute queued steps
+/// - `eval_step()` / `eval_output()` - run to completion
+/// - `get_queued_steps()` / `get_step_result()` - inspect state
+pub struct DebugExecutor {
+    /// Per-item execution state (dependency tracking, results)
+    state: ItemState,
+    /// State store for persistence
+    state_store: Arc<dyn StateStore>,
+    /// Stepflow executor for plugin access
+    executor: Arc<StepflowExecutor>,
+    /// The workflow being executed
+    flow: Arc<Flow>,
+    /// Execution context for this session
+    context: ExecutionContext,
+}
+
+impl DebugExecutor {
+    /// Create a new debug executor for the given workflow and input.
+    ///
+    /// This is for fresh executions. Use [`resume()`](Self::resume) to resume
+    /// an existing execution from the state store.
+    pub fn new(
+        executor: Arc<StepflowExecutor>,
+        flow: Arc<Flow>,
+        flow_id: BlobId,
+        run_id: Uuid,
+        input: ValueRef,
+        state_store: Arc<dyn StateStore>,
+        variables: Option<HashMap<String, ValueRef>>,
+    ) -> Result<Self> {
+        // Validate the workflow before execution
+        let diagnostics =
+            stepflow_analysis::validate(&flow).change_context(ExecutionError::AnalysisError)?;
+
+        if diagnostics.has_fatal() {
+            let fatal = diagnostics.num_fatal;
+            let error = diagnostics.num_error;
+            return Err(
+                error_stack::report!(ExecutionError::AnalysisError).attach_printable(format!(
+                    "Workflow validation failed with {fatal} fatal and {error} error diagnostics"
+                )),
+            );
+        }
+
+        // Create ItemState with workflow input and variables
+        let state = ItemState::new(
+            flow.clone(),
+            input,
+            Arc::new(variables.unwrap_or_default()),
+        );
+
+        // Create workflow-aware execution context
+        let context = ExecutionContext::for_workflow_with_flow(
```

Comment:
Naming? `for_workflow_with_flow`? Is this still needed? Can it be cleaned up after #472?